File: copymove/src/main/scala/farjs/copymove/MoveProcess.scala
package farjs.copymove

import farjs.filelist.FileListActions
import farjs.filelist.FileListActions.FileListTaskAction
import farjs.filelist.api.FileListItem
import farjs.ui.Dispatch
import farjs.ui.popup._
import farjs.ui.task.Task
import farjs.ui.theme.Theme
import scommons.nodejs.{FS, path}
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Future, Promise}
import scala.scalajs.js
import scala.util.{Failure, Success}

case class MoveProcessProps(dispatch: Dispatch,
                            actions: FileListActions,
                            fromPath: String,
                            items: Seq[(FileListItem, String)],
                            toPath: String,
                            onTopItem: FileListItem => Unit,
                            onDone: () => Unit)

object MoveProcess extends FunctionComponent[MoveProcessProps] {

  private[copymove] var statusPopupComp: ReactClass = StatusPopup
  private[copymove] var messageBoxComp: ReactClass = MessageBox
  private[copymove] var fs: FS = scommons.nodejs.fs

  private case class MoveState(currItem: String = "",
                               existing: Option[String] = None)

  protected def render(compProps: Props): ReactElement = {
    val (state, setState) = useStateUpdater(() => MoveState())
    val inProgress = useRef(false)
    val existsPromise = useRef(Promise.successful[Boolean](true))
    val askWhenExists = useRef(true)
    val currTheme = Theme.useTheme()
    val props = compProps.wrapped

    def moveItems(): Unit = {
      val resultF = props.items.foldLeft(Future.successful(true)) { case (resF, (currItem, toName)) =>
        resF.flatMap {
          case true if inProgress.current =>
            setState(_.copy(currItem = currItem.name))
            val oldPath = path.join(props.fromPath, currItem.name)
            val newPath = path.join(props.toPath, toName)
            val exists = !currItem.isDir && fs.existsSync(newPath)

            if (exists && askWhenExists.current) {
              setState(_.copy(existing = Some(newPath)))
              existsPromise.current = Promise[Boolean]()
            }
            existsPromise.current.future.flatMap { overwrite =>
              if (!exists || overwrite) {
                fs.rename(oldPath, newPath).map { _ =>
                  props.onTopItem(currItem)
                  inProgress.current
                }
              } else Future.successful(inProgress.current)
            }
          case res => Future.successful(res)
        }
      }
      resultF.onComplete {
        case Success(_) => props.onDone()
        case Failure(_) =>
          props.onDone()
          props.dispatch(FileListTaskAction(Task("Moving items", resultF)))
      }
    }

    def onExistsAction(overwrite: Boolean,
                       all: Boolean = false,
                       cancel: Boolean = false): js.Function0[Unit] = { () =>
      
      setState(_.copy(existing = None))
      askWhenExists.current = !all
      inProgress.current = !cancel
      existsPromise.current.trySuccess(overwrite)
    }

    useLayoutEffect({ () =>
      // start
      inProgress.current = true
      moveItems()
      ()
    }, Nil)

    <.>()(
      <(statusPopupComp)(^.plain := StatusPopupProps(
        text = s"Moving item\n${state.currItem}",
        title = "Move",
        onClose = { () =>
          // stop
          inProgress.current = false
        }: js.Function0[Unit]
      ))(),

      state.existing.map { existing =>
        <(messageBoxComp)(^.plain := MessageBoxProps(
          title = "Warning",
          message = s"File already exists.\nDo you want to overwrite it's content?\n\n$existing",
          actions = js.Array(
            MessageBoxAction("Overwrite", onExistsAction(overwrite = true)),
            MessageBoxAction("All", onExistsAction(overwrite = true, all = true)),
            MessageBoxAction("Skip", onExistsAction(overwrite = false)),
            MessageBoxAction("Skip all", onExistsAction(overwrite = false, all = true)),
            MessageBoxAction(
              label = "Cancel",
              onAction = onExistsAction(overwrite = false, all = true, cancel = true),
              triggeredOnClose = true
            )
          ),
          style = currTheme.popup.error
        ))()
      }
    )
  }
}

File: copymove/MoveProcess.mjs
/**
 * @import { Dispatch } from "@farjs/filelist/FileListData.mjs"
 * @import { FileListItem } from "@farjs/filelist/api/FileListItem.mjs"
 */
import path from "path";
import nodeFs from "fs";
import fsPromises from "fs/promises";
import React, { useLayoutEffect, useRef, useState } from "react";
import Task from "@farjs/ui/task/Task.mjs";
import TaskAction from "@farjs/ui/task/TaskAction.mjs";
import MessageBox from "@farjs/ui/popup/MessageBox.mjs";
import StatusPopup from "@farjs/ui/popup/StatusPopup.mjs";
import FileListActions from "@farjs/filelist/FileListActions.mjs";
import Theme from "@farjs/ui/theme/Theme.mjs";
import MessageBoxAction from "@farjs/ui/popup/MessageBoxAction.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  existsSync(path: string): boolean;
 *  rename(oldPath: string, newPath: string): Promise<void>;
 * }} FS
 */

/**
 * @typedef {{
 *  readonly dispatch: Dispatch;
 *  readonly actions: FileListActions;
 *  readonly fromPath: string;
 *  readonly items: readonly {
 *    readonly item: FileListItem;
 *    readonly toName: string;
 *  }[];
 *  readonly toPath: string;
 *  onTopItem(topItem: FileListItem): void;
 *  onDone(): void;
 * }} MoveProcessProps
 */

/**
 * @typedef {{
 *  readonly currItem: string;
 *  readonly existing?: string;
 * }} MoveState
 */

/**
 * @param {MoveProcessProps} props
 */
const MoveProcess = (props) => {
  const { statusPopupComp, messageBoxComp } = MoveProcess;

  const [state, setState] = useState(
    () => /** @type {MoveState} */ ({ currItem: "" })
  );
  const inProgress = useRef(false);
  const existsPromise = useRef({
    /** @type {Promise<boolean>} */
    p: Promise.resolve(true),
    /** @type {(v: boolean) => void} */
    resolve: () => {},
  });
  const askWhenExists = useRef(true);
  const currTheme = Theme.useTheme();

  const moveItems = () => {
    const resultP = props.items.reduce(async (resP, cpItem) => {
      const res = await resP;
      const { item: currItem, toName } = cpItem;
      if (res && inProgress.current) {
        setState((s) => {
          return { ...s, currItem: currItem.name };
        });
        const oldPath = path.join(props.fromPath, currItem.name);
        const newPath = path.join(props.toPath, toName);
        const exists = !currItem.isDir && MoveProcess.fs.existsSync(newPath);

        if (exists && askWhenExists.current) {
          setState((s) => {
            return { ...s, existing: newPath };
          });
          /** @type {(v: boolean) => void} */
          let resolve = () => {};
          const p = new Promise((res) => (resolve = res));
          existsPromise.current = { p, resolve };
        }

        const overwrite = await existsPromise.current.p;
        if (!exists || overwrite) {
          await MoveProcess.fs.rename(oldPath, newPath);
          props.onTopItem(currItem);
          return inProgress.current;
        }
        return inProgress.current;
      }
      return res;
    }, Promise.resolve(true));

    resultP.then(
      () => props.onDone(),
      () => {
        props.onDone();
        props.dispatch(TaskAction(Task("Moving items", resultP)));
      }
    );
  };

  /** @type {(overwrite: boolean, all?: boolean, cancel?: boolean) => () => void} */
  function onExistsAction(overwrite, all = false, cancel = false) {
    return () => {
      setState((s) => {
        return { ...s, existing: undefined };
      });
      askWhenExists.current = !all;
      inProgress.current = !cancel;
      existsPromise.current.resolve(overwrite);
    };
  }

  useLayoutEffect(() => {
    // start
    inProgress.current = true;
    moveItems();
  }, []);

  const existing = state.existing;

  return h(
    React.Fragment,
    null,
    h(statusPopupComp, {
      text: `Moving item\n${state.currItem}`,
      title: "Move",
      onClose: () => {
        // stop
        inProgress.current = false;
      },
    }),

    existing !== undefined
      ? h(messageBoxComp, {
          title: "Warning",
          message: `File already exists.\nDo you want to overwrite it's content?\n\n${existing}`,
          actions: [
            MessageBoxAction("Overwrite")(onExistsAction(true)),
            MessageBoxAction("All")(onExistsAction(true, true)),
            MessageBoxAction("Skip")(onExistsAction(false)),
            MessageBoxAction("Skip all")(onExistsAction(false, true)),
            MessageBoxAction("Cancel", true)(onExistsAction(false, true, true)),
          ],
          style: currTheme.popup.error,
        })
      : null
  );
};

MoveProcess.displayName = "MoveProcess";
MoveProcess.statusPopupComp = StatusPopup;
MoveProcess.messageBoxComp = MessageBox;
/** @type {FS} */
MoveProcess.fs = { ...nodeFs, ...fsPromises };

export default MoveProcess;
